src/runner_manager.py

```diff
@@ -0,0 +1,358 @@
+
+        logger.info("Installing charm dependencies.")
+
+        # Binding for snap, apt, and lxd init commands are not available so subprocess.run used.
+        execute_command(["/usr/bin/apt", "remove", "-qy", "lxd", "lxd-client"], check=False)
+        execute_command(["/usr/bin/snap", "install", "lxd", "--channel=latest/stable"])
+        execute_command(["/usr/bin/snap", "refresh", "lxd", "--channel=latest/stable"])
+        execute_command(["/snap/bin/lxd", "waitready"])
+        execute_command(["/snap/bin/lxd", "init", "--auto"])
+        execute_command(["/usr/bin/chmod", "a+wr", "/var/snap/lxd/common/lxd/unix.socket"])
+        execute_command(["/snap/bin/lxc", "network", "set", "lxdbr0", "ipv6.address", "none"])
+        execute_command(
+            [
+                "/usr/bin/apt",
+                "install",
+                "-qy",
+                "cpu-checker",
+                "libvirt-clients",
+                "libvirt-daemon-driver-qemu",
+            ],
+        )
+        cls.runner_bin_path.parent.mkdir(parents=True, exist_ok=True)
+
+        logger.info("Finished installing charm dependencies.")
+
+    def get_latest_runner_bin_url(self, os_name: str = "linux", arch_name: str = "x64") -> str:
+        """Get the URL for the latest runner binary.
+
+        The runner binary URL changes when a new version is available.
+
+        Returns:
+            The URL to the latest runner binary.
+        """
+
+        runner_bins: RunnerApplicationList = []
+        if isinstance(self.path, GitHubRepo):
+            runner_bins = self._github.actions.list_runner_applications_for_repo(
+                owner=self.path.owner, repo=self.path.repo
+            )
+        elif isinstance(self.path, GitHubOrg):
+            runner_bins = self._github.actions.list_runner_application_for_org(org=self.path.org)
+
+        logger.debug("Response of runner binary list: %s", runner_bins)
+
+        for bin in runner_bins:
+            if bin.os == os_name and bin.architecture == arch_name:
+                return bin.download_url
+
+        raise RunnerBinaryError(
+            f"Unable to download GitHub runner binary for {os_name} {arch_name}"
+        )
+
+    @retry(tries=10, delay=15, max_delay=60, backoff=1.5)
+    def update_runner_bin(self, download_url: str) -> None:
+        """Download a runner file, replacing the current copy.
+
+        Args:
+            download_url (str): URL to download the runner binary.
+
+        TODO:
+            Deletion first cause no runner binary on download failure. See if this cause problems.
+        """
+        logger.info("Downloading runner binary from: %s", download_url)
+
+        # Remove any existing runner bin file
+        if self.runner_bin_path.exists():
+            self.runner_bin_path.unlink()
+        # Download the new file
+        response = self.session.get(download_url)
+
+        logger.info("Download of runner binary return status code: %i", response.status_code)
```

Comment:
Should we log the `sha` or some other identifying information as well? Also, should we check the download has not been messed with? Not sure if we can verify it against a signature, for example